{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": [],
      "gpuType": "A100",
      "machine_shape": "hm",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Noble-Lab/HiCFoundation/blob/main/HiCFoundation.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# HiCFoundation: a generalizable Hi-C foundation model for chromatin architecture, single-cell and multi-omics analysis across species"
      ],
      "metadata": {
        "id": "nEDektTYN-5v"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<a href=\"https://github.com/marktext/marktext/releases/latest\">\n",
        "   <img src=\"https://img.shields.io/badge/HiCFoundation-v1.0.0-green\">\n",
        "   <img src=\"https://img.shields.io/badge/platform-Linux%20%7C%20Mac%20-green\">\n",
        "   <img src=\"https://img.shields.io/badge/Language-python3-green\">\n",
        "   <img src=\"https://img.shields.io/badge/dependencies-tested-green\">\n",
        "</a>  \n",
        "\n",
        "HiCFoundation is a generalizable Hi-C foundation model for chromatin architecture, single-cell and multi-omics analysis across species.\n",
        "\n",
        "Copyright (C) 2024 Xiao Wang, Yuanyuan Zhang, Suhita Ray, Anupama Jha, Tangqi Fang, Shengqi Hang, Sergei Doulatov, William Stafford Noble, and Sheng Wang\n",
        "\n",
        "License: Apache License 2.0\n",
        "\n",
        "Contact:  Sergei Doulatov (doulatov@uw.edu) & William Stafford Noble (wnoble@uw.edu) & Sheng Wang (swang@cs.washington.edu)\n",
        "\n",
        "For technical problems or questions, please reach to Xiao Wang (wang3702@uw.edu) and Yuanyuan Zhang (zhang038@purdue.edu).\n",
        "\n",
        "**We strongly suggest to use Google Chrome for HiCFoundation Colab. Other browsers such as Safari may raise errors when uploading or downloading files.**\n",
        "\n",
        "If you are using other browsers, disabling tracking protection may help resolve the errors when uploading or downloading files.\n",
        "\n",
        "For more details, see **<a href=\"#Instructions\">Instructions</a>** of the notebook and checkout the **[HiFoundation GitHub](https://github.com/Noble-Lab/HiCFoundation)**. If you use HiCFoundation, please cite it: **<a href=\"#Citation\">Citation</a>**."
      ],
      "metadata": {
        "id": "DF8V39buPEyX"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Overall Protocol\n",
        "1) Pre-training stage: the model is trained in a self-supervised fashion on massive quantities of unlabeled Hi-C data. The model takes masked Hi-C submatrices as input, optimizing for the reconstruction of the full submatrix.\n",
        "\n",
        "2) Fine-tuning stage: the model is fine-tuned and tested for diverse downstream applications, including genome architecture analysis, multi-species analysis, neutrophil differentiation analysis, multi-omics analysis and single-cell analysis.\n",
        "\n",
        "\n",
        "<p align=\"center\">\n",
        "  <img src=\"https://github.com/Noble-Lab/HiCFoundation/raw/main/imgs/framework_github.png\" alt=\"HiCFoundation framework\" width=\"70%\">\n",
        "</p>"
      ],
      "metadata": {
        "id": "MqWYOHjaVTve"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Instructions <a name=\"Instructions\"></a>\n",
        "## Steps\n",
        "1. Connect to a **gpu machine** by clicking the right top button **\"connect\"** in the notebook, then we can run HiCFoundation under GPU support. We strongly recommend to connect with **A100** with **high memory**.(This is included as paid option).<br>\n",
        "You can also connect to local runtime or GCE VM on Google cloud.\n",
        "2. Click the left running button in <a href=\"#Dependency\">Install Dependencies</a> to install dependencies.\n",
        "3. Upload .hic/.cool/.pairs file in <a href=\"#file\">Input file</a> for inference.\n",
        "4. Specify the Parameters in <a href=\"#Param\">Parameters</a>. Please make sure to click the left button to configure the parameters after your modification. If you used the default one, please also run it.\n",
        "5. Running HiCFoundation by by clicking the left running button in <a href=\"#Running\">Run HiCFoundation</a>.\n",
        "6. Download HiCFoundation output. Please click the left running button in <a href=\"#Download\">Download</a> to download the zip files. <br>\n",
        "\n",
        "PS: For **reproducibility task**, please run 3-6 for interested Hi-C maps to get their corresponding embedding files in .pkl format. Keep them in your local computer, then you can run <a href=\"https://github.com/Noble-Lab/HiCFoundation/blob/main/Reproducibility.ipynb\">Repro Colab</a> to calculate the similarity by uploading the saved embedding files here."
      ],
      "metadata": {
        "id": "ydQCdUbqW8M5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title <a name=\"Dependency\">Install dependencies</a>\n",
        "#@markdown Please make sure the notebook is already connected to **GPU**, HiCFoundation needs GPU support to run.\n",
        "#@markdown <br> It will restart the kernel and throughout some error, but it is fine, please just ignore it and let it run.\n",
        "\n",
        "%%shell\n",
        "cd /content\n",
        "git clone https://github.com/Noble-Lab/HiCFoundation --quiet\n",
        "cd HiCFoundation\n",
        "\n",
        "pip install -q condacolab\n",
        "python3 -c \"import condacolab;condacolab.install()\"\n",
        "eval \"$(conda shell.bash hook)\"\n",
        "conda env create -f /content/HiCFoundation/environment_notorch.yml\n",
        "conda activate HiCFoundation\n",
        "pip install torch==1.8.1+cu111 torchvision==0.9.1+cu111 torchaudio==0.8.1 -f https://download.pytorch.org/whl/torch_stable.html\n",
        "pip install timm==0.3.2\n",
        "cd /content/HiCFoundation\n",
        "cd hicfoundation_model\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_pretrain.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_reproducibility.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_loop.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_loop_lc.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_resolution.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_epigenomic.pth.tar\n",
        "wget https://huggingface.co/wang3702/hicfoundation_models/resolve/main/hicfoundation_schic.pth.tar\n",
        "cd ..\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "1wyxpaF9RffO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title  <a name=\"file\">Input file</a>\n",
        "from google.colab import files\n",
        "import os\n",
        "import os.path\n",
        "import re\n",
        "import hashlib\n",
        "import random\n",
        "import string\n",
        "from google.colab import drive\n",
        "\n",
        "from datetime import datetime\n",
        "# Get the current date and time\n",
        "current_datetime = datetime.now()\n",
        "# Convert to string in desired format\n",
        "current_datetime_str = current_datetime.strftime(\"%Y-%m-%d-%H-%M-%S\")\n",
        "\n",
        "\n",
        "rand_letters = string.ascii_lowercase\n",
        "rand_letters = ''.join(random.choice(rand_letters) for i in range(20))\n",
        "#@markdown ## Option 1: download from database <br>\n",
        "#@markdown Instead of uploading, you can also specify the link here to automatically download maps from ENCODE and other servers.\n",
        "#@markdown Example: https://www.encodeproject.org/files/ENCFF689CUX/@@download/ENCFF689CUX.hic\n",
        "download_link = '' #@param {type:\"string\"}\n",
        "output_dir=\"/content/\"\n",
        "if download_link!='':\n",
        "  root_dir = os.getcwd()\n",
        "  upload_dir = os.path.join(root_dir,rand_letters)\n",
        "  if not os.path.exists(upload_dir):\n",
        "    os.mkdir(upload_dir)\n",
        "  os.chdir(upload_dir)\n",
        "  os.system(\"wget %s\"%download_link)\n",
        "  parse_link=download_link.split(\"/\")[-1]\n",
        "  hic_input_path = os.path.join(upload_dir,parse_link)\n",
        "  os.chdir(root_dir)\n",
        "  output_dir=\"/content/hicfoundation_output/\"\n",
        "\n",
        "else:\n",
        "  #@markdown ## Option 2: Use files in Google Drive\n",
        "  #@markdown If you wanted to use google drive, please input the file path in your google drive\n",
        "  #@markdown <br> This is recommended when your file is larger than 1GB.\n",
        "  #@markdown <br> Please authenticate google colab to access your google drive, whic will pop up a window for your authentication.\n",
        "  #@markdown <br> **The colab is also private to you and Google, developers can not see your data, so privacy is guaranteed.**\n",
        "  #@markdown <br> The output is directly saved to your google drive, you can directly download and analyze the data on your Google Drive.\n",
        "  google_drive_file = \"\" #@param {type:\"string\"}\n",
        "\n",
        "  if google_drive_file==\"\":\n",
        "    #@markdown ## Option 3: Upload from your local file system\n",
        "    #@markdown This is not recommended for big files. Please only consider this when input file is small (<=100 MB).\n",
        "    print(\"Please uploading your input files\")\n",
        "    os.chdir(\"/content/\")\n",
        "    root_dir = os.getcwd()\n",
        "    upload_dir = os.path.join(root_dir,rand_letters)\n",
        "    if not os.path.exists(upload_dir):\n",
        "      os.mkdir(upload_dir)\n",
        "    os.chdir(upload_dir)\n",
        "    map_input = files.upload()\n",
        "    for fn in map_input.keys():\n",
        "      print('User uploaded file \"{name}\" with length {length} bytes'.format(\n",
        "        name=fn, length=len(map_input[fn])))\n",
        "      hic_input_path = os.path.abspath(fn)\n",
        "      print(\"The input save to %s\"%hic_input_path)\n",
        "    os.chdir(root_dir)\n",
        "    output_dir=\"/content/hicfoundation_output/\"\n",
        "\n",
        "  else:\n",
        "\n",
        "    #loading the data from google drive\n",
        "\n",
        "    drive.mount('/content/drive')\n",
        "    hic_input_path = \"/content/drive/My Drive/\"+google_drive_file\n",
        "    if not os.path.exists(hic_input_path):\n",
        "      print(f\"The specified path {hic_input_path} did not contain input file.\")\n",
        "      print(f\"You may consider simply upload your file to the rood directory of your Google drive. Then you only need to input the file name in the field.\")\n",
        "      print(\"If you still encounter problems, please contact us!\")\n",
        "      exit()\n",
        "    print(f\"Google drive path is available at {hic_input_path}\")\n",
        "    output_dir=f\"/content/drive/My Drive/hicfoundation_output_{current_datetime_str}\"\n",
        "os.makedirs(output_dir,exist_ok=True)\n",
        "print(f\"HiCFoundation output will be redirected to {output_dir}\")\n"
      ],
      "metadata": {
        "collapsed": true,
        "cellView": "form",
        "id": "nf32iXaLW7WO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "l45Q20LpPcq5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "#@title  <a name=\"Param\">Configure parameters</a>\n",
        "#@markdown We supported six different tasks, please make selection based on your need.\n",
        "#@markdown - Reproducibility analysis: HiCFoundation will generate embeddings of the input Hi-C, and the submatrix embeddings can be used to compare across biological replicates and non-replicates.\n",
        "#@markdown - Chromatin loop detection: HiCFoudation will generate the loop detection of the input Hi-C in .bedpe format.\n",
        "#@markdown - Resolution enhancement: HiCFoundation will generate enhanced Hi-C map given the input Hi-C.\n",
        "#@markdown - Epigenomic assay profiling: HiCFoundation will generate corressponding epigenomic assays in .bigWig format given the input Hi-C.\n",
        "#@markdown - Single-cell Hi-C enhancement: HiCFoundation will generate the enhanced scHi-C given the input siHi-C.\n",
        "#@markdown - Hi-C Embedding: Generates different levels of embeddings for Hi-C:\n",
        "#@markdown  - patch level embdding: an embedding vector corresponds to a 16*16 patch space at specified resolution.\n",
        "#@markdown  - submatrix level embedding: an embedding vector corresponds to the specified submatrix at specified resolution.\n",
        "#@markdown  - chromosome level embedding: embedding vectors correspond to different chromosomes at specified resolution.\n",
        "#@markdown  - genome wide embedding: an embedding vector corresponds to the input Hi-C at specified resolution.\n",
        "\n",
        "task = \"Hi-C Embedding\" # @param [\"Reproducibility analysis\", \"Loop detection\", \"Resolution enhancement\",\"Epigenomic profiling\",\"Single-cell Hi-C\",\"Hi-C Embedding\"]\n",
        "task_name=task\n",
        "task_list=[\"Reproducibility analysis\", \"Loop detection\", \"Resolution enhancement\",\"Epigenomic profiling\",\"Single-cell Hi-C\",\"Hi-C Embedding\"]\n",
        "task = task_list.index(task)+1\n",
        "root_dir=\"/content/HiCFoundation/hicfoundation_model/\"\n",
        "if task==1:\n",
        "  model_path=root_dir+\"hicfoundation_reproducibility.pth.tar\"\n",
        "elif task==2:\n",
        "  model_path=root_dir+\"hicfoundation_loop.pth.tar\"\n",
        "elif task==3:\n",
        "  model_path=root_dir+\"hicfoundation_resolution.pth.tar\"\n",
        "elif task==4:\n",
        "  model_path=root_dir+\"hicfoundation_epigenomic.pth.tar\"\n",
        "elif task==5:\n",
        "  model_path=root_dir+\"hicfoundation_schic.pth.tar\"\n",
        "else:\n",
        "  model_path=root_dir+\"hicfoundation_pretrain.pth.tar\"\n",
        "#configure input size\n",
        "#@markdown Configure the input submatirx size of HiCFoundation. <br>\n",
        "#@markdown For \"Reproducibility analysis\", \"Loop detection\", \"Resolution enhancement\",and \"Single-cell Hi-C\" task, 224$\\times$224 is suggested input size.\n",
        "#@markdown <br> For \"Epigenomic Profiling\", the suggested input size is 128$\\times$4000.\n",
        "#@markdown <br> For \"Hi-C Embedding\", please chooce the submatrix size that you are interested. Please make sure the numbers are multiplies of 16.\n",
        "\n",
        "submat_height = 128 #@param {type:\"number\"}\n",
        "submat_width = 4000 #@param {type:\"number\"}\n",
        "#@markdown Configure scanning stride for the input Hi-C matrix. The default should applies to most settings.\n",
        "stride = 32 #@param {type:\"number\"}\n",
        "#@markdown Configure scanning boundary for the input Hi-C matrix. Default 0 indicates scanning across the diagonal, should work for most tasks. You can adjust based on your need.\n",
        "scan_boundary = 0 #@param {type:\"number\"}\n",
        "#@markdown Configure the batch size for HiCFoundation inference.\n",
        "batch_size = 1 #@param {type:\"number\"}\n",
        "#@markdown Configure resolution for inference:\n",
        "#@markdown - Chromatin loop detection/Resolution enhancement: 10000 (10kb).\n",
        "#@markdown - Reproducibility analysis: 25000 (25kb)\n",
        "#@markdown - Epigenomic assay profilling: 1000 (1kb)\n",
        "#@markdown - Single-cell Hi-C analysis: 1000000 (1Mb)\n",
        "resolution = 10000 #@param {type:\"number\"}\n",
        "\n",
        "#@markdown (Only applicable to \"Resolution enhancement\") Configure genome id or the path to chromsize files.\n",
        "genome_id = \"hg38\" #@param {type:\"string\"}\n",
        "#@markdown (Only applicable to \"Hi-C embedding\") Configure the embedding depth: default: 0 (encoder output embeddings). You can also specify k from 1 to 8 to indicate the output of k-th layer of the pre-trained decoder.\n",
        "embed_depth = 0 #@param {type:\"number\"}\n",
        "#@markdown (Only applicable to \"Loop detection\") Configure if this is a low-coverage Hi-C (like <=100Mb reads for human genome).\n",
        "low_coverage_input = False # @param {type:\"boolean\"}\n",
        "print(\"Configured task for inference: %s\"%task_name)\n",
        "if task==2 and low_coverage_input:\n",
        "  model_path=root_dir+\"hicfoundation_loop_lc.pth.tar\"\n",
        "  print(\"Low coverage input for loop detection, used model %s\"%model_path)\n",
        "else:\n",
        "  print(\"Configured HiCFoundation model path %s\"%model_path)\n",
        "command_line=f\"python3 inference.py --input '{hic_input_path}' --batch_size {batch_size} --resolution {resolution} \\\n",
        "  --task {task} --input_row_size {submat_height} --input_col_size {submat_width} \\\n",
        "  --stride {stride} --bound {scan_boundary} --model_path '{model_path}' \\\n",
        "  --output '{output_dir}'\"\n",
        "print(\"Configure command:\",command_line)\n",
        "#write it to a running file\n",
        "with open(\"/content/HiCFoundation/run.sh\",\"w\") as file:\n",
        "  file.write(\"%s\"%command_line)"
      ],
      "metadata": {
        "id": "ul4yLfwbQzpp",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title <a name=\"Running\">Run HiCFoundation</a>\n",
        "#@markdown Please allow >30 minutes to finish the output. The epigenomic assay may take longer because of its finer resolution.\n",
        "#@markdown <br>Our running time is correlated to the size of input Hi-C.\n",
        "#@markdown <br>If your Hi-C contact map is too big, please run locally with our [GitHub](https://github.com/Noble-Lab/HiCFoundation).\n",
        "#@markdown <br>If you don't have GPU resources, please make contact with us and we are happy to run it for you.\n",
        "\n",
        "%%shell\n",
        "cd /content/HiCFoundation\n",
        "git pull origin main\n",
        "eval \"$(conda shell.bash hook)\"\n",
        "conda activate HiCFoundation\n",
        "bash /content/HiCFoundation/run.sh\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "o8h41jtc_XuA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title <a name=\"Download\">Download Output</a>\n",
        "#@markdown If you want to download all output files in .zip file, please select this.\n",
        "#@markdown Otherwise, it will be downloaded in .tar.gz file.\n",
        "from google.colab import files\n",
        "import os, tarfile\n",
        "import shutil\n",
        "import zipfile\n",
        "\n",
        "zip_format = True #@param {type:\"boolean\"}\n",
        "download_path = \"/content/download/\"\n",
        "os.makedirs(download_path,exist_ok=True)\n",
        "print(\"Copying files from %s\"%output_dir)\n",
        "listfiles=[x for x in os.listdir(output_dir) if \"HiCFoundation\" in x]\n",
        "for item in listfiles:\n",
        "  src_file = os.path.join(output_dir,item)\n",
        "  dest_file = os.path.join(download_path,item)\n",
        "  shutil.copy(src_file,dest_file)\n",
        "if zip_format:\n",
        "  tar_path = \"/content/HiCFoundation_output.zip\"\n",
        "else:\n",
        "  tar_path = \"/content/HiCFoundation_output.tar.gz\"\n",
        "def zip_file(tar_path,src_dir):\n",
        "    zip_name = tar_path\n",
        "    z = zipfile.ZipFile(zip_name,'w',zipfile.ZIP_DEFLATED)\n",
        "    for dirpath, dirnames, filenames in os.walk(src_dir):\n",
        "        fpath = dirpath.replace(src_dir,'')\n",
        "        fpath = fpath and fpath + os.sep or ''\n",
        "        for filename in filenames:\n",
        "            z.write(os.path.join(dirpath, filename),fpath+filename)\n",
        "            print ('==Compress Success!==',filename)\n",
        "    z.close()\n",
        "\n",
        "def make_targz(output_filename, source_dir):\n",
        "    \"\"\"\n",
        "    :param output_filename:\n",
        "    :param source_dir:\n",
        "    :return: bool\n",
        "    \"\"\"\n",
        "    try:\n",
        "        with tarfile.open(output_filename, \"w:gz\") as tar:\n",
        "            tar.add(source_dir, arcname=os.path.basename(source_dir))\n",
        "\n",
        "        return True\n",
        "    except Exception as e:\n",
        "        print(e)\n",
        "        return False\n",
        "if zip_format:\n",
        "  zip_file(tar_path,download_path)\n",
        "else:\n",
        "  make_targz(tar_path,download_path)\n",
        "files.download(tar_path)\n",
        "\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "v9JL7CZuT4Kh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Citation: <a name=\"Citation\"></a>\n",
        "\n",
        "Xiao Wang, Yuanyuan Zhang, Suhita Ray, Anupama Jha, Tangqi Fang, Shengqi Hang, Sergei Doulatov, William Stafford Noble, and Sheng Wang. \"A generalizable Hi-C foundation model for chromatin architecture, single-cell and multi-omics analysis across species.\" bioRxiv (2024).\n",
        "<a href=\"https://www.biorxiv.org/content/10.1101/2024.12.16.628821v1\">Paper</a>\n",
        "```\n",
        "@article{wang2024generalizable,\n",
        "  title={A generalizable Hi-C foundation model for chromatin architecture, single-cell and multi-omics analysis across species},\n",
        "  author={Wang, Xiao and Zhang, Yuanyuan and Ray, Suhita and Jha, Anupama and Fang, Tangqi and Hang, Shengqi and Doulatov, Sergei and Noble, William Stafford and Wang, Sheng},\n",
        "  journal={bioRxiv},\n",
        "  year={2024}\n",
        "}\n",
        "```"
      ],
      "metadata": {
        "id": "FyNShuKvDiBB"
      }
    }
  ]
}